### PR TITLE
Replace a few instances of PtrToStructure with more efficient marshalling

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/libraries/Common/src/Interop/Windows/Crypt32/Interop.FindOidInfo.cs [hunk 1]
```diff
@@ -75,28 +75,28 @@ internal static CRYPT_OID_INFO FindOidInfo(CryptOidInfoKeyType keyType, string k
                 if (!OidGroupWillNotUseActiveDirectory(group))
                 {
                     OidGroup localGroup = group | CRYPT_OID_DISABLE_SEARCH_DS_FLAG;
-                    IntPtr localOidInfo = CryptFindOIDInfo(keyType, rawKey, localGroup);
-                    if (localOidInfo != IntPtr.Zero)
+                    CRYPT_OID_INFO* localOidInfo = CryptFindOIDInfo(keyType, rawKey, localGroup);
+                    if (localOidInfo != null)
                     {
-                        return Marshal.PtrToStructure<CRYPT_OID_INFO>(localOidInfo);
+                        return *(CRYPT_OID_INFO*)localOidInfo;
```
Comment (huoyaoyuan) [inline src/libraries/Common/src/Interop/Windows/Crypt32/Interop.FindOidInfo.cs:81]: Why is a cast required in these returns?

Comment (jkotas) [reply src/libraries/Common/src/Interop/Windows/Crypt32/Interop.FindOidInfo.cs]: Not needed. I forgot to delete them when I have fixed the PInvoke signature to return the proper type.

**Comment by rzikm** [comment]
Out of curiosity, what does `Marshal.PtrToStructure<T>` do that it makes it less efficient than unsafe casting and dereferencing the raw pointer?

**Comment by Fabi** [comment]
> Out of curiosity, what does `Marshal.PtrToStructure<T>` do that it makes it less efficient than unsafe casting and dereferencing the raw pointer?

Most likely calls like Activator.CreateInstance etc? This change also made these parts reflection free compatible

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemUnicastIPAddressInformation.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ internal sealed class SystemUnicastIPAddressInformation : UnicastIPAddressInform
         private readonly uint _preferredLifetime;
```
Comment (am11) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemUnicastIPAddressInformation.cs:20]: From this list:
<details>
  <summary><code>git grep -l PtrToStructure ':/*.cs' ':/!*[Tt]est*'</code></summary>

```sh
src/coreclr/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs
src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs
src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.NativeAot.cs
src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs
src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedInteropStubManager.cs
src/coreclr/tools/aot/ILLink.Shared/TrimAnalysis/IntrinsicId.cs
src/coreclr/tools/aot/ILLink.Shared/TrimAnalysis/Intrinsics.cs
src/libraries/Common/src/System/Net/NetworkInformation/HostInformationPal.Windows.cs
src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs
src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComTypeEnumDesc.cs
src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/IDispatchComObject.cs
src/libraries/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs
src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcUtils.cs
src/libraries/System.Data.OleDb/src/DbPropSet.cs
src/libraries/System.Data.OleDb/src/OleDbConnectionInternal.cs
src/libraries/System.Data.OleDb/src/OleDbDataReader.cs
src/libraries/System.Data.OleDb/src/PropertyInfoSet.cs
src/libraries/System.Data.OleDb/src/System/Data/ProviderBase/DbBuffer.cs
src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs
src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/AD/SidList.cs
src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/AuthZSet.cs
src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/SAM/SAMStoreCtx.cs
src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs
src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/Interop/LdapPal.Linux.cs
src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/BerConverter.cs
src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs
src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs
src/libraries/System.DirectoryServices/src/Interop/AdsValueHelper2.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/ActiveDirectorySite.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/AttributeMetaData.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryContext.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServer.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Domain.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DomainController.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Exception.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Forest.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/ForestTrustRelationshipInformation.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Locator.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/ReplicationCursorCollection.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/ReplicationFailure.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/ReplicationNeighbor.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/ReplicationOperation.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs
src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs
src/libraries/System.Drawing.Common/src/System/Drawing/Font.cs
src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.cs
src/libraries/System.Drawing.Common/src/System/Drawing/Printing/PageSettings.cs
src/libraries/System.Drawing.Common/src/System/Drawing/Printing/PrinterSettings.cs
src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs
src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs
src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs
src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs
src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs
src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs
src/libraries/System.Speech/src/AudioFormat/AudioFormatConverter.cs
src/libraries/System.Speech/src/Internal/SapiInterop/SapiRecoInterop.cs
src/libraries/System.Speech/src/Internal/StreamMarshaler.cs
src/libraries/System.Speech/src/Result/RecognitionResult.cs
src/libraries/System.Speech/src/Result/RecognizedPhrase.cs
src/libraries/System.Windows.Extensions/src/System/Media/SoundPlayer.cs
src/mono/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs
```
</details>

looks like usages in `HostInformationPal.Windows.cs` and `Ping.Windows.cs` can also be avoided.

**Comment by jkotas** [comment]
> > Out of curiosity, what does Marshal.PtrToStructure<T> do that it makes it less efficient than unsafe casting and dereferencing the raw pointer?

> Most likely calls like Activator.CreateInstance etc? This change also made these parts reflection free compatible

It also needs to lookup the marshalling strategy for the type.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemUnicastIPAddressInformation.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemUnicastIPAddressInformation.cs:20]: https://github.com/dotnet/runtime/pull/70900

**Pull request merged**